constraintlayout/compose/src/main/java/androidx/constraintlayout/compose/MotionSceneScope.kt

```diff
@@ -0,0 +1,259 @@
+    }
+
+    override fun setDebugName(name: String?) {
+        // Do nothing
+    }
+
+    override fun getTransition(str: String?): String {
+        // Do nothing
+        return ""
+    }
+
+    override fun getConstraintSet(index: Int): String {
+        // Do nothing
+        return ""
+    }
+
+    override fun getConstraintSetInstance(name: String): ConstraintSet? {
+        return constraintSetsByName[name]
+    }
+
+    override fun getTransitionInstance(name: String): Transition? {
+        return transitionsByName[name]
+    }
+}
+
+/**
+ * Scope used by the MotionScene DSL.
+ *
+ * &nbsp
+ *
+ * Define new [ConstraintSet]s and [Transition]s within this scope using [constraintSet] and
+ * [transition] respectively.
+ *
+ * Alternatively, you may add existing objects to this scope using [addConstraintSet] and
+ * [addTransition].
+ *
+ * &nbsp
+ *
+ * The [defaultTransition] **should always be set**. It defines the initial state of the layout and
+ * works as a fallback for undefined `from -> to` transitions.
+ */
+@ExperimentalMotionApi
+class MotionSceneScope internal constructor(private val dpToPixel: CorePixelDp) {
+    /**
+     * Count of generated ConstraintSet & Transition names.
+     */
+    private var generatedCount = 0
+
+    /**
+     * Returns a new unique name. Should be used when the user does not provide a specific name
+     * for their ConstraintSets/Transitions.
+     */
+    private fun nextName() = UNDEFINED_NAME_PREFIX + generatedCount++
+
+    internal var constraintSetsByName = HashMap<String, ConstraintSet>()
+    internal var transitionsByName = HashMap<String, Transition>()
+
+
+    internal fun reset() {
+        generatedCount = 0
+        constraintSetsByName.clear()
+        transitionsByName.clear()
+    }
+
+    /**
+     * Defines the default [Transition], where the [from] and [to] [ConstraintSet]s will be the
+     * initial start and end states of the layout.
+     *
+     * The default [Transition] will also be applied when the combination of `from` and `to`
+     * ConstraintSets was not defined by a [transition] call.
+     *
+     * This [Transition] is required to initialize [MotionLayout].
+     */
+    fun defaultTransition(
+        from: ConstraintSetRef,
+        to: ConstraintSetRef,
+        transitionContent: TransitionScope.() -> Unit = { }
+    ) {
+        transition("default", from, to, transitionContent)
+    }
+
+    /**
+     * Adds a [ConstraintSet] defined by [constraintSetContent]. A [name] may be provided and it can
+     * be used on MotionLayout calls that request a ConstraintSet name.
+     *
+     * Returns a [ConstraintSetRef] object representing this ConstraintSet, which may be used as a
+     * parameter of [transition].
+     *
+     * If not null, inherits the changes applied by [extendConstraintSet]
+     */
+    fun constraintSet(
+        name: String? = null,
+        extendConstraintSet: ConstraintSetRef? = null,
```

Comment:
nit: I was wondering if it would be a bit more clear (this constraintSet is created based on the this constraintSet) if we rename it to something like baseConstraintSet or derivedConstraintSet? It may not be necessary though.

Reply:
I reused the name we use in a similar ConstraintSet method.
```
fun ConstraintSet(
    extendConstraintSet: ConstraintSet,
    description: ConstraintSetScope.() -> Unit
)
```
Updated Javadoc with a similar wording as that function.